lib/global_uid/allocator.rb

```diff
@@ -0,0 +1,66 @@
+module GlobalUid
+  class Allocator
+    attr_reader :allocations, :max_size, :incrementing_by, :connection
+
+    def initialize(max_size: 5, incrementing_by:, connection:)
+      @allocations = []
+      @max_size = max_size
+      @incrementing_by = incrementing_by
+      @connection = connection
+      validate_connection_increment(connection)
+    end
+
+    def add(table)
+      identifier = connection.insert("REPLACE INTO #{table} (stub) VALUES ('a')")
+      increment!(identifier)
+    end
+
+    def add_many(table, count:)
+      increment_by = validate_connection_increment(connection)
+
+      start_id = connection.insert("REPLACE INTO #{table} (stub) VALUES " + (["('a')"] * count).join(','))
+      identifiers = start_id.step(start_id + (count-1) * increment_by, increment_by).to_a
+      identifiers.each { |identifier| increment!(identifier) }
+      identifiers
+    end
+
+    private
+
+    def increment!(identifier)
+      allocations.shift if allocations.size >= max_size
+      allocations << identifier
+
+      begin
+        unless valid_increment?
+          raise InvalidIncrementException, \
+            "Recently allocated IDs: #{allocations} on '#{connection.current_database}'"
+        end
+        identifier
+      rescue InvalidIncrementException => e
+        raise unless GlobalUid::Base.global_uid_options[:suppress_increment_exceptions]
+        GlobalUid::Base.notify(e, e.message)
+        identifier
+      end
+    end
+
+    def valid_increment?
+      allocations[1..-1].all? do |identifier|
+        (identifier - allocations[0]) % incrementing_by == 0
+      end
+    end
+
+    def validate_connection_increment(connection)
```

Comment:
Since the `connection` being passed in is always available by the `#connection` reader method, I think we should just remove this parameter.